### PR TITLE
Cpo

### PR DESCRIPTION
**Pull request by MJHamar**
What I did so far:
`train_generator.py`: a rough draft of the training pipeline.
- `ExtraArguments`: args that are specific to the training pipeline, incomplete
- `peft_finetune` method:
  - loads the model and the tokenizer
  - loads the dataset
  - TODO: build the collator, it should be a `DPODataCollatorWithPadding`
  - TODO: build a `transformers.Trainer`
  - TODO: train
  - TODO: save model
  - etc.
- `DSPyConfig`: dspy specific arguments, if we choose to tune using dspy as well
- `dspy_finetune` method

`cpo_dataset.py`: defines methods for building and loading the dataset for CPO tuning
- `load_llmt_dataset`: aims to mimic the data loading logic they did in the ALMA repo
  - loads a json file, which contains documents, example documents for few-shot prompting, generated and gt queries and corresponding scores. It performs the 2-out-of-3 selection and returns an iterable dataset. All this is TODO
- `build_llmt_dataset`: describes the whole pipeline for building the dataset that `load_llmt_dataset` expects, including generating the queries by both the teacher and the student models.
  - implemented through a series of sub-functions for readability
  - Builds a JSON file in the following format:
    ```
        {
            "dataset_name": str -- name of the dataset,
            "doc_ids": List[str] -- list of document IDs,
            "data": {
                {doc_id}: {
                    "example_ids": List[str], -- list of example document IDs
                    "example_texts": List[str], -- list of example document strings
                    "example_query_ids": List[str], -- list of example query IDs
                    "example_queries": List[str], -- list of example queries

                    "target_doc_id": str, -- the target document ID
                    "target_doc_text": str, -- the target document string
                    "ref_query_id": str, -- id of the ground truth query
                    "ref_query": str, -- the ground truth query string
                    "teacher_query": str, -- the teacher query
                    "student_query": str, -- the student query

                    "ref_score": float, -- the ground truth score
                    "teacher_score": float, -- the teacher score
                    "student_score": float, -- the student score
                }
            }
        }
        ```
  - Two things are still missing:
    1. We need to decide how we want to select the example document-query pairs for the prompt
    2. We need to implement the `generate_queries` function. I would like to use it for generating queries for both the big and the small model.

NOTES:
I'll remove the ALMA submodule, we don't need it, just keeping it for reference rn.

### CODE REVIEW

**Comment by MJHamar** [comment]
almost ready, something is still messed up with the way we perprocess the data. Will fix tomorrow

File: inpars/cpo_finetune.py [hunk 1]
```diff
@@ -39,6 +39,13 @@ def __post_init__(self):
         self.tokenizer = AutoTokenizer.from_pretrained(self.model_name_or_path)
         if self.tokenizer.pad_token_id is None:
             self.tokenizer.add_special_tokens({'pad_token': '[PAD]'})
```
Comment (MJHamar) [inline inpars/cpo_finetune.py]: we no longer need this line do we

File: inpars/cpo_dataset.py
Comment (JesseBrouw) [inline inpars/cpo_dataset.py:1]: looks great!

File: inpars/cpo_finetune.py (hunk 1 above)
Comment (m-krastev) [reply inpars/cpo_finetune.py]: Why do you think so?

Comment (MJHamar) [reply inpars/cpo_finetune.py]: In the following line we set the pad_token property directly.
i also tried running without this line and had no issues

File: inpars/cpo_dataset.py [hunk 2]
```diff
@@ -0,0 +1,730 @@
...
+                            batched=True,
+                            batch_size=1,
+                            num_proc=data_args.preprocessing_num_workers,
+                            load_from_cache_file=not data_args.overwrite_cache,
+                            desc="Running CPO preprocessing",
+                        )
+                    else:
+                        train_dataset = train_dataset.map(
+                            cpo_prompt_function, batched=True, batch_size=1
+                        )
```
Comment (m-krastev) [inline inpars/cpo_dataset.py]: What's the logic behind these lines? @MJHamar

File: inpars/cpo_dataset.py [hunk 3]
```diff
@@ -0,0 +1,730 @@
...
+            )
+            lowest_score_index = (
+                TEACHER_OUTPUT_KEY if example[TEACHER_SCORE_KEY] < example[REF_SCORE_KEY] else REF_OUTPUT_KEY
+            )
+
+        # Assigning the corresponding sentences
+        highest_score_sentence = example[highest_score_index]
+        lowest_score_sentence = example[lowest_score_index]
+
+        return highest_score_sentence, lowest_score_sentence
```
Comment (m-krastev) [inline inpars/cpo_dataset.py:76]: I need someone to check that the logic behind the unrolled loop is right. As a side note, the unrolled loop results in 2x improvement (from ~600ns to ~300ns). Though I still wonder whether the nanosecond range matters here.

File: requirements.txt [hunk 4]
```diff
@@ -144,3 +144,5 @@ zlib-state==0.1.9
 requests
 bm25s==0.1.0
 pytest==8.3.3
+trl==0.12.1
+bitsandbytes>=0.44.0
```
Comment (MJHamar) [inline requirements.txt:148]: Don't forget to also add `unsloth`

**Pull request merged**